presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/GroupArrayAggregationState.java

```diff
@@ -106,15 +138,14 @@ public long getEstimatedSize()
                 nextPosition.sizeOf() +
                 valueBlocksRetainedSizeInBytes +
                 // valueBlocksRetainedSizeInBytes doesn't contain the current block builder
-                currentBlockBuilder.getRetainedSizeInBytes();
+                currentPageBuilder.getRetainedSizeInBytes();
     }
 
-    @Override
-    public void add(Block block, int position)
+    protected void add(Consumer<PageBuilder> consumer)
```

Comment:
I agree.  I'll just quickly document why I chose this approach, and allow me to propose an alternative solution for your feedback.

I believe, though the JDK documentation specifies that `Optional` is a value-based class, this line was added to the documentation to allow for the JVM to implement Optional as a value type for future releases, and currently Optional behaves like other boxed types.  That means I would be worried of the object allocation overhead when it's not empty.

I could make this a nullable type, but my concern was that it wasn't obvious from the type what to do when it's null, and it seemed like a code smell.  I wanted to move code that operated on one value in the unary subclass, and the code that operated on two values in the binary subclass.  Note that the above code block won't work, because the position may differ between `block1` and `block2`.  We would also need another parameter for `position2`.

The compromise I considered and implemented was to pass the Consumer, so that subclasses could mutate the PageBuilder directly.  In terms of performance, initially this should perform like a method call, and after it's JIT compiled it should be inlined.  It is true that inside this method we don't really know what exactly adds to this PageBuilder, that happens in the consumer and there is a level of indirection for that.  But the subclasses are all in the same class, so it's just a matter of navigating to the abstract add method.

I can consider three approaches to make it more obvious.  One is to rename this method from `add` to something like `prepareNewEntry`, or some method name that shows the intent of this method, which is to perform the recordkeeping necessary to add an entry.  And the methods in the subclasses remain `add`.  That way when someone looks for the `add` method, they'll just see the subtype's methods, and it will be pretty clear that the call inside them is a recordkeeping call specific to this implementation.

Another is to just make `block2` nullable, and add another int parameter for the second position.  That should be fine too, I can just annotate the `block2` as `@Nullable`, but I wanted to avoid relying on this sort of if null checking as it seemed to have a code smell from my perspective.

A third option is to just fork this code between the unary and binary types.  I don't think this would be too bad, given the growing type gymnastics needed to prevent the fork.  But I personally don't think option 1 is that bad.  But please let me know what your feedback is on this.

As a followup, after our discussion, I will add a benchmark that will try to determine if there's any overhead associated with boxing the Optional type, or with using the `Consumer`.

Reply:
Microbenchmarks don't show a start performance impact of one or the other, and I did confirm the lambda will be inlined ([see gist here](https://gist.github.com/tdcmeehan/d63863563201ce3286f64141ca8b450a)).  I discussed offline with @wenleix and will be going with a variant of option 1, which is to rename the base `add` method ass `prepareAdd`, which will perform recordkeeping for linked list updates, and the base classes will directly add on the `currentPageBuilder` via a new method.  This also won't require the `Consumer` so it may be a bit more straightforward reading it as well.